記事検索

myokoym opened the pull request:
レイアウトの右サイドバーに記事検索の検索ボックスを追加しました。

機能としては以下です。
- 記事のタイトル、本文から検索
- 記事のタイトルを表示
- 関係者で絞り込み可能

関係者の名前や説明での検索は今はできていませんが、必要なら実装可能です。

Close #185 

TODO:
- [x] 公開後に公開停止した記事が検索で引っかかってしまわないようにする https://github.com/bm-sms/daimon-news-multi-tenant/pull/288/files#r54970392
  - 普通にARのモデルにマッピングすると順序が失われるけど、最新日付順でよいのでは
- [x] 検索ボタンの多言語化（日本語化）
- [x] 著者情報検索のレビュー対応（不要なインデックスを作らない）
  - [x] https://github.com/bm-sms/daimon-news-multi-tenant/commit/c15ca8bc8211a61f25b223838cb623390e6ebba2#commitcomment-16527161

別PR:
- [ ] テキスト抽出（マークダウンの記号で引っかからないようにする）
- [ ] https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#discussion_r54832043
- [ ] 著者情報検索のレビュー対応（著者情報フルネームの順位を改善）
  - [ ] https://github.com/bm-sms/daimon-news-multi-tenant/commit/c15ca8bc8211a61f25b223838cb623390e6ebba2#commitcomment-16527154
- [ ] 著者情報検索のレビュー対応（著者情報の重み付け改善）
  - [ ] https://github.com/bm-sms/daimon-news-multi-tenant/commit/c15ca8bc8211a61f25b223838cb623390e6ebba2#commitcomment-16557270
- [ ] strong parameterでpermitするの減らす？ https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#discussion_r55311282
- [ ] 検索結果のページネーションは不要？ https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#issuecomment-193617934
- [ ] 検索結果ページでは検索ボックスはサイドバーではなく検索結果のところに表示？ https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#issuecomment-193616616
- [ ] 検索後に検索ボックスが空になってしまう
- [ ] サムネイルを表示する

kou committed:
Support importing data to Groonga

kou committed:
Implement related posts by Groonga

kou committed:
Import public_id to Groonga table

It's for generating URL from data in Groonga.

kou committed:
Add search page

It has only simple full text search by title and text.

kou committed:
Only show posts for the current site

kou committed:
Load author and category details

kou committed:
Support empty keywords

kou committed:
Support drilldown by author

kou committed:
Show drilldown links by author

kou committed:
Use single quote for String

kou committed:
Add missing published check

kou committed:
Fix drilldown links on no query case

Commit:
Use app/groonga instead of app/models

Commit:
Put search form on sidebar

Commit:
author -> participant

Commit:
Move label position

Commit:
Prevent error for preview in such as /editor route

Commit:
Fix typos

credit      ->
participant

Commit:
Add missing block local variable

Commit:
Remove needless receiver

Commit:
Add missing "s"

Commit:
Author -> Participant

Commit:
Save category.name and participant.name

myokoym commented:
ほぼreview readyですが、以下のTODOが残っています。

  * 検索結果が0件だった場合に「見つかりませんでした」メッセージを表示する
  * ドリルダウンのリンクはあるけど絞り込めない

Commit:
Fix search condition

Commit:
Move search code to SearchController from ApplicationController

myokoym commented:
> 検索結果が0件だった場合に「見つかりませんでした」メッセージを表示する

必要なら後から追加することにします。

okkez commented:
>>    検索結果が0件だった場合に「見つかりませんでした」メッセージを表示する
>
> 必要なら後から追加することにします。

入れられない理由がないなら、入れた方がいいと思います。
検索結果がないのに、何も表示されないと何が悪いのかわからなくて使う人が戸惑うからです。

mtsmfm commented:
著者もひっかかってくれると嬉しそうですねぃ

okkez commented:
テキスト抽出も、ちゃんとやった方がいい。

mtsmfm commented:
「キーワード」の検索結果
みたいなのがあると、「これ何で検索したんだっけ」ってならないで嬉しいかもです!

okkez commented:
`app/groonga` にしたのはなんでだろう？

okkez commented:
* [ ] 検索結果0件の表示
* [ ] 検索したキーワードも表示

この2点をやってリリース :shipit: したい。

mtsmfm commented:
:+1:

okkez committed:
Display search result when nothing is found

okkez committed:
Display total count for search result

okkez commented:
検索結果100件出るのは多い気がした。

okkez commented:
> 検索結果100件出るのは多い気がした。

ドリルダウンが付けば、問題ないか？

最低限はやったので、これでお願いします :pray:

okkez committed:
EmptyPosts should be empty

kou commented on app/groonga/post_searcher.rb:
```diff
@@ -22,4 +22,89 @@ def related_post_ids(post, limit: 5)
+    retried = false
+    begin
+      sorted_posts = posts.paginate([['_score', :desc]],
+                                    page: (page || 1).to_i,
```
こことこの下って`page`と`size`をそのまま使っちゃダメなんでしたっけ。

kou commented on app/views/search/_form.html.haml:
```diff
@@ -0,0 +1,3 @@
+= form_for @query, url: search_path, method: 'get' do |f|
+  = f.search_field :keywords
+  = f.submit 'search'
```
`participant_id`もhiddenで入れておかないとキーワードを再編集すると関係者での絞り込みが消えてしまいます。

myokoym replied:
あー、どうなんでしょう。「aaa」で検索→「関係者A」で絞り込み→「bbb」で検索としたときに、
「bbb」の検索結果が「関係者A」で絞りこまれているべき、という話ですよね。
再検索するときは絞り込み結果をリセットしてもいい気もしています。

kou commented on groonga/init.rb:
```diff
@@ -28,12 +28,13 @@
 if Post.table_exists?
   indexer = PostIndexer.new
-  Post.find_each do |post|
+  Post.includes(:category).find_each do |post|
```
関連するレコードは事前に読み込んでおかないとN+1問題なやつになって起動が遅くなるんじゃないかと心配しています。なので、`credits`とかも読んでおいた方がよいと思います。

okkez committed:
Display score as data attribute

Score is needed by developer.

okkez committed:
Simplify code

Because `page` and `size` are already integer.

okkez committed:
Avoid error when delete record on Groonga

See #291

mtsmfm commented on app/groonga/post_indexer.rb:
```diff
@@ -23,12 +25,13 @@ def add(post)
+               public_id: post.public_id)
   end
 
   def remove(post)
```
#291 で話したのですが、管理画面からの変更の同期がんばっても dyno 間で同期取れないのでやめちゃいましょ!

== remove とか要らなくなりますね

myokoym replied:
3c0b689b47251f279f189b560ec50509a015ae82 の一環ですね。使わないなら消してしまいますか。
Heroku以外で動かしたい人がいたら使われるかも？と思ったけど、その時に考えればいいか。

mtsmfm commented on app/groonga/post_searcher.rb:
```diff
@@ -22,4 +22,89 @@ def related_post_ids(post, limit: 5)
+      return result_set
+    end
+
+    page = (page || 1).to_i
```
(page || 1) はデフォルト引数にすればやらなくて済みます?

kou replied:
呼ぶ側で`search(page: nil)`とできなくなるので不便な気がしますけどねぇ。

mtsmfm replied:
あー、 `search(page: params[:page])` と書きたい?

mtsmfm replied:
あったあった

https://github.com/bm-sms/daimon-news-multi-tenant/pull/288/files#diff-f1d36687c4c649ebbb025139ed60ff37R4

なるほどです :bulb:

myokoym commented:
別々のdynoでadd/removeされるとエラーになるので、記事ごとの追加/削除時の同期はしない。
一日1回の再起動で同期されればよしとする。

mtsmfm commented on app/groonga/post_searcher.rb:
```diff
@@ -22,4 +22,89 @@ def related_post_ids(post, limit: 5)
+      sorted_posts = posts.paginate([['_score', :desc]],
+                                    page: page,
+                                    size: size)
+    rescue Groonga::TooSmallPage, Groonga::TooLargePage
```
(あんま groonga のことよくわからないので聞いちゃいますが) ここで例外補足する必要があるのってなんででしょうか?

kou replied:
無効な`page/size`を指定すると例外があがるからです。（ってそういうことじゃない？）

mtsmfm replied:
- 正常系の処理に例外が使われていないか
- ここで握り潰していい処理なのか

が気になったのでした!

> 無効なpage/sizeを指定すると

クライアントエラーを返すようになってた方がよかったりしませんか?

kou replied:
> クライアントエラーを返すようになってた方がよかったりしませんか?

それでもいいですが、たぶん、不便ですよ。

無効な`page/size`になるときってだいたいユーザーが悪いんじゃないんですよ。昔貼ったリンクが無効になってしまって、それを辿ったときとかに発生します。

そのとき、エラーになるよりは同じ検索条件で1ページ目の方がユーザーに優しいんじゃないかなぁと私は思っています。

mtsmfm replied:
ふむー。

ならば、こんな感じで retry やめるとどうでしょう!
フラグで retry の制御つらいなーと思ったのでした

``` ruby
# controller
# XXX もうちょいいい感じの名前...
post_searcher.search_with_invalid_page_handling(@query, page: params[:page])
```

``` ruby
class PostSearcher
  def search_with_invalid_page_handling(query, page: nil, size: nil)
    search(query, page: page, size: size)
  rescue Groonga::TooSmallPage, Groonga::TooLargePage
    search(query, page: 1, size: size)
  end
end
```

kou replied:
このプロジェクトでは`retry`を避ける（メンテナンスする人たちが`retry`を避けたい）ということであれば、`retry`を使わない実装でもよいと思います！

mtsmfm replied:
retry が読みやすければ retry でいいんですが、僕の思いとしては以下の感じです:
- フラグと引数書き換えによる retry の制御が読みにくい
- `search` という大きな名前にエラーハンドリングが入っているのがもやっとする

kou replied:
あ、はい、メンテナンスする人たちがメンテナンスしやすい実装でよいと思っています！

kou replied:
あ、ちなみに、私が書き換えるなら`paginate`関連でメソッドを分けると思います。

myokoym replied:
こんな感じにしてみました https://github.com/bm-sms/daimon-news-multi-tenant/commit/865635f22862a07b58461f066e3b106704ba03b7

okkez commented:
モデルのコールバックを消したので、既に記事ごとの同期は止まっています、というのをお伝えしておきます。

mtsmfm commented on app/models/post.rb:
```diff
@@ -13,16 +13,6 @@ def with_ordered
 
   before_save :assign_public_id
 
-  after_save do |post|
```
この辺りの変更は master に入ってるので rebase してもらえればと

myokoym replied:
はいー

mtsmfm commented on app/controllers/application_controller.rb:
```diff
@@ -24,4 +26,9 @@ def routing_error!
   end
+
+  def prepare_search
+    @query = Query.new(params[:query])
```
ActiveModel だから怒られない気がしますが、strong parameter 使っておきたいような

`params.require(:query).permit(:keywords, ...)`

myokoym replied:
登録フォームではないですしエラーにする必要はないかと思ったので、必須チェックなしで対応してみました。 a2e42e0f5e8859a83a969695c785b2f36afc5a7c

mtsmfm commented on app/views/search/search.html.haml:
```diff
@@ -0,0 +1,28 @@
+    - if @query.participant_id.present?
+      関係者：
+      = Participant.find(@query.participant_id).name
+      = link_to content_tag(:span, '', class: 'fa fa-remove'), @query.params(remove: [:participant_id])
```
`@query` が Hash っぽいインタフェースだと読みやすかったりするんですかねぇ?

`@query.except(:participant_id)`
`@query.merge(participant_id: participant._key)`

あるいは `@query.params.merge()` とか

mtsmfm replied:
主なモチベーションとしては:
- `Query#params` の `add` `remove` オプションを覚えなくても、 Hash の語彙でいける気がした
- `Query#params` で send 使うのやめたい

mtsmfm replied:
なんかこんな形にできる予感がしています

``` ruby
class Query
def attributes
  {keywords: keywords, site_id: site_id, participant_id: participant_id}
end

def params
  {query: attributes}
end

def present?
  values = attributes.except(:site_id).values
  values.present? && values.all?(&:present?)
end
end
```

kou replied:
`Hash#except`はActive Supportが生やしているメソッドなので`Hash`の語彙と言われても（私は）ピンとこないんですが、Railsアプリケーションですし、メンテナンスする人たちがそっちの方が読みやすそうならそういう感じの実装でよいと思います！

tricknotes replied:
> なんかこんな形にできる予感がしています

form オブジェクトをサポートする gem を導入すると、やりたいことを実現しつつ自分たちで厄介なコードを書かなくて済むので考えることが減ってメンテナンスしやすくなるのでは、と思いました。

tricknotes commented on app/assets/stylesheets/themes/default/_search.sass:
```diff
@@ -0,0 +1,2 @@
+span.score
```
`span` タグであることって重要なんですか？
（わたしなら `.score` と指定するな、と思いました）

okkez replied:
score自体画面に出さないことにしたので重要でないです。

okkez replied:
削除漏れです。。。

myokoym replied:
削除しました

tricknotes replied:
> 削除しました

このファイル自体、もう不要になってませんか？
（たぶん、rebase ミスってる気配を感じます）

myokoym replied:
あー、.scssに変換してからファイル自体不要なことに気づいて削除したのですが、変換前の.sassが残っていました。。。
ありがとうございます。消しておきます。

tricknotes commented on app/controllers/application_controller.rb:
```diff
@@ -8,6 +8,8 @@ class ApplicationController < ActionController::Base
     before_action :authorize_user!, if: :current_user
   end
 
+  before_action :prepare_search
```
`before_action` で暗黙的にインスタンス変数を設定するのって、どうなんでしょうねぇ…。
しかもこれって ApplicationController なので、これを継承するすべてのコントローラで `@query` というインスタンス変数が定義されてしまうのって、ちょっと影響が大きすぎるような…？

myokoym replied:
確かにそうかもしれません。代替手段としては以下のどれかでしょうか？
- 各コントローラーのアクション毎（:indexと:show？）に定義する
- インスタンス変数を使わなくてもいいようにする
- https://github.com/apotonick/cells などでビューごと切り出す

myokoym replied:
お安く、各コントローラーのアクション毎に定義するようにしました。

tricknotes replied:
I think:

> `before_action` で暗黙的にインスタンス変数を設定するのって、どうなんでしょうねぇ…。

tricknotes replied:
なぜかというと、インスタンス変数の初期化は action 自体の責務だと思うからです。
`before_action` という名前に期待することは、「アクションの前処理」つまりaction を処理するための事前条件のチェック（ログインしているか・権限があるか、など）だとわたしは考えているからです。

tricknotes replied:
他に `before_action` が適切だなと思う場面は、action とは関係なくを事前にやっておきたい処理（例えばログ出力とか）がある場合かなと考えています。

myokoym replied:
>  `before_action` で暗黙的にインスタンス変数を設定するのって、どうなんでしょうねぇ…。

すみません、別問題でしたね。
layoutで使っている部分なのでbefore_actionでいいかと思っていたのですが、確かにaction内で定義するのが普通ですか。修正します。

tricknotes replied:
layout でインスタンス変数を使いたくなること自体が何か怪しい気がするので、もういっそ cells とかそれに類する gem を入れてみてもいいのかな、と言う気がしますね…！

kou replied:
検索画面以外のときに、サイドバーにあるフォームに値が入っていることってあるんですか？

myokoym replied:
> 検索画面以外のときに、サイドバーにあるフォームに値が入っていることってあるんですか？

ないと思います。インスタンス変数を不要にできるのではないか、という話ですか？
`form_for`で使っていたので生成していましたが、不要にできるならそうしたほうがいいですか。

``` ruby
= form_for @query, url: search_path, method: 'get' do |f|
```

myokoym replied:
`form_tag`を使うようにして、layoutでインスタンス変数を使わなくて済むようにしました。

myokoym commented on app/views/search/search.html.haml:
```diff
@@ -0,0 +1,28 @@
+- else
+  %div
+    - if @query.participant_id.present?
+      関係者：
```
ドリルダウン結果のラベルは「絞り込み」とかのほうがいいかも

Commit:
Remove needless style

span.score has been deleted from template.

Commit:
Remove a unused method

PostIndexer#remove has been unused because an error occurred if it
runs on other dyno that is not added dyno.

Commit:
Use form_tag instead of form_for not to use query object for search form in layout

Commit:
Don't use "retry" to simplify for out of range on pagination

Commit:
Use strong parameters for search

Commit:
Fix dirilldown form label

myokoym commented:
マージできるところまではきたと思います。
TODOは以下です。

* [ ] 著者もひっかかってくれると嬉しそうですねぃ
* [ ] テキスト抽出も、ちゃんとやった方がいい。
* [ ] https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#discussion_r54832043

できれば対応したいですが、別PRのほうがよいかもしれません。

tricknotes commented on app/controllers/search_controller.rb:
```diff
@@ -0,0 +1,12 @@
+    @result_set = searcher.search_with_fallback_to_first_page(@query, page: search_params[:page])
+  end
+
+  def search_params
```
`private`

myokoym replied:
追加しました :bow:

tricknotes commented on app/groonga/query.rb:
```diff
@@ -0,0 +1,28 @@
+  attr_accessor :participant_id
+
+  def present?
+    keywords.present? ||
```
パッと見て「`keywords` か `participant_id` が存在している」というのがわかりづらかったです(改行位置のせいかな…)

たとえばこんなのはどうでしょうか？

``` ruby
[keyword, participant_id].any?(&:present?)
```

kou replied:
私には、このシンプルな述語2個のORのために、配列を作って `any?` を使って `&:` も使うのはやりすぎな気がしますけど。。。

Railsアプリケーションだとそういうもんなんですかねぇ。

（念のため書いておくと、メンテナンスする人たちが読みやすい方を選べばいいと私は思っています。）

tricknotes replied:
なるほどなるほどー！
確かに余分な処理をしていると言われればその通りですね…。

cosmo0920 replied:
よく分かってないんですけど、僕は今のままが読みやすいと感じました。 hoge_clause `||` で改行されていたら次にまだ続くんだろうなーと言う感じで読めました。`||` は二項演算子なので`||`の右側が欠けても左側が欠けても意味が通らなくなってしまう（型が合わなくなってしまう）ので。。

tricknotes replied:
>  `||` で改行されていたら次にまだ続くんだろうなー

なるほどー、今までこの感覚は自分にはなかったので、これからコードを見る時の良いヒントになりそうです！

tricknotes commented on db/seeds/development/credits.seeds.rb:
```diff
@@ -1,11 +1,10 @@
+  site1.posts.published.order_by_recently.limit(10).each_with_index do |post, i|
     post.credits.create!(
-      participant: participant,
+      participant: site1.participants[i % 2],
```
ここの `2` って、seed で用意している件数でしょうか？
`2` だけだとあまりにもマジックナンバーで意図がわかりづらいな、と思いました。

もし「全部の participants をいい感じにローテートしたい」というのでよいとすると、こんなのはどうでしょうか？

``` ruby
participants = site1.participants.cycle

# ...

site1.posts.published.order_by_recently.limit(10).each do |post|
  post.credits.create!(
    participant: participants.next
  )
end
```

myokoym replied:
これは雑すぎますね…検証用のコードを後で改善しようと思ってそのままになっていました…。
「全部の participants をいい感じにローテートしたい」ということだったので、使わせてもらいました！
（Enumlatorのこういう使い方はあまりしたことがなかったので勉強になります :memo: ）

mtsmfm replied:
おー、 cycle なるほど

tricknotes commented on groonga/init.rb:
```diff
@@ -28,12 +28,13 @@
 if Post.table_exists?
   indexer = PostIndexer.new
-  Post.find_each do |post|
+  Post.includes(:category, credits: [:participant]).find_each do |post|
```
このコードは master で見たような気がするので、 rebase お願いします！

myokoym replied:
rebaseしました！

tricknotes commented:
> できれば対応したいですが、別PRのほうがよいかもしれません。

この合理的な理由ってありますか？
（せめているわけではないんですが）これらの TODO も含めて「ひとつの変更」のように思えたので、あえて別の PR に分ける理由がわたしにはよくわかっていないです。

Commit:
Increase credits seeds

myokoym commented:
>
この合理的な理由ってありますか？
（責めているわけではないんですが）これらの TODO も含めて「ひとつの変更」のように思えたので、あえて別の PR に分ける理由がわたしにはよくわかっていないです。

ご指摘ありがとうございます！すみません、説明不足でした。
私が考えていたのは、まず最低限の実装で入れて、使ってもらいながら改善していく
ほうが効率がいいのかも、というところでした。

（↓でそういう方針になったと思っていました）
https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#issuecomment-191003049
https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#issuecomment-191023784

が、たしかに「ひとつの変更」と言えますし、やることが決まっているのなら
一緒に入れたほうがよさそうですね。

Commit:
Support post.participants for full-text-search target

Commit:
Add missing "private" for strong parameters

Commit:
Add missing permitted parameters "utf8" and "commit" for search form

myokoym commented:
> 著者もひっかかってくれると嬉しそうですねぃ

https://github.com/bm-sms/daimon-news-multi-tenant/commit/c15ca8bc8211a61f25b223838cb623390e6ebba2 で対応しました。
@kou レビューをお願いできますでしょうか？

Commit:
Decrease pagination size for search (100 -> 50)

myokoym commented:
> 検索結果100件出るのは多い気がした。

確かに多すぎると思ったので50件にしました。 https://github.com/bm-sms/daimon-news-multi-tenant/commit/cedb36bf8eb20ddc378dfe7acdd80173846f0908

mtsmfm commented on app/groonga/post_searcher.rb:
```diff
@@ -24,4 +24,89 @@ def related_post_ids(post, limit: 5)
+    posts = @posts.select do |record|
+      conditions = []
+      conditions << (record.published_at > 0)
+      conditions << (record.published_at <= Time.now)
```
groonga が返す published_at は古い可能性があるので、反響を受けて慌てて公開停止にした場合に検索結果では見えるけれどリンクを踏むと見えない、ということになりそうですね。

記事関連と同じく、 `Post.published` を噛ませる必要がありそうです。

https://github.com/bm-sms/daimon-news-multi-tenant/issues/291#issuecomment-191066678

tricknotes commented:
わたしは不安な部分にはテストを書いていきたいのですが、この検索機能には不安はありませんか？

mtsmfm commented:
> わたしは不安な部分にはテストを書いていきたいのですが、この検索機能には不安はありませんか？

僕は不安なのでテスト足しときます

mtsmfm commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  test 'only posts of current site' do
+    visit '/search?query[keywords]=contents'
```
ユーザと同じように操作した方が、`IntegrationTest` としての価値は高いと思います

visit '/' して fill_in して click_on 'search' するといいんじゃないでしょうか

mtsmfm commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,27 @@
+  test 'only posts of current site' do
+    visit '/search?query[keywords]=contents'
+    within('main') do
+      assert_equal '「contents」を含む記事は1件見つかりました。', find('p').text
```
「1 件見つかりました」だと、post 1 と post 2 のどっちがひっかかったかわからないですよね
タイトルで検証するのはどうでしょうか

tricknotes commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,27 @@
+    site1 = create(:site)
+    site2 = create(:site)
+    body = 'contents...'
+    post1 = create(:post, site: site1, title: 'post1!', body: body)
```
xxx1 と xxx2 ってどう違うんでしょうか？
というのがコードからは読み取れませんでした。

それがどういうデータなのか、変数名からそれが読み取れると頭の負担が減って読みやすいコードになると思います。
わたしだったら `post_should_be_found` のような名前をつけるなと思いました。

kou commented:
> > できれば対応したいですが、別PRのほうがよいかもしれません。
> 
> この合理的な理由ってありますか？
> （責めているわけではないんですが）これらの TODO も含めて「ひとつの変更」のように思えたので、あえて別の PR に分ける理由がわたしにはよくわかっていないです。

私なら、すんごい便利でカンペキな記事検索を実現する！なら「ひとつの変更」でよいと思いますが、今はないので記事検索をできるようにしよう！ならわけた方がいいなぁと思います。で、私ならすんごい便利でカンペキな記事検索を実現する！はそもそもひとつのタスクにしないなぁと思います。

それとは別に、こんなに大きくて長生きするpull requestがあることがあんまりよい状況じゃないなぁと思います。（というのが私の感覚ですが、それが一般的かどうかはわからないです。）

tricknotes commented:
> こんなに大きくて長生きするpull requestがあることがあんまりよい状況じゃないなぁと思います。

（一般的なかどうかはわからないですが）完全に同意です！
この PR に関して言うと diff は全然大きくないと思うので、なにか進め方自体を改善できそうだなと感じてます。

> 私ならすんごい便利でカンペキな記事検索を実現する！はそもそもひとつのタスクにしないなぁと思います。

これも完全に同意です。
小さくて整合性のとれた状態で少しづつ開発を進められそうなら、その単位でマージできる方がスムーズに開発していけると感じています。
（少なくともわたしがコメントした段階では整合性が取れていない状態に見えました。）

Commit:
test: add Groonga database setup/teardown methods

kou commented on app/controllers/search_controller.rb:
```diff
@@ -0,0 +1,14 @@
+  private
+
+  def search_params
+    params.permit(:page, :utf8, :commit, query: [:keywords, :participant_id, :site_id])
```
`:utf8`とか`:commit`とかも入れないといけないのがもやっとしました。
けど、strong parametersを使うっていうのはそういうことなんですよねぇ。。。

tricknotes replied:
奇遇ですね、わたしももやっとします…！
使う必要がないパラメータは `permit` する必要がないように感じますねぃ

tricknotes replied:
> strong parametersを使う

もともとの意図は、「期待しないパラメータが `Query.new` のコンストラクタに渡らないようにしたい」(ref https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#discussion_r54831734) なので、意図と違って余分なパラメータを permit してしまっているように見えますね。

tricknotes replied:
@myokoym 
ここはこれから直しますか？

myokoym replied:
> `:utf8`とか`:commit`とかも入れないといけないのがもやっとしました。
> けど、strong parametersを使うっていうのはそういうことなんですよねぇ。。。
> 
> もともとの意図は、「期待しないパラメータが `Query.new` のコンストラクタに渡らないようにしたい」(ref https://github.com/bm-sms/daimon-news-multi-tenant/pull/288#discussion_r54831734) なので、意図と違って余分なパラメータを permit してしまっているように見えますね。

すみません、コメントできていませんでした。

`Query.new`のコンストラクタに渡らないようにするという意味では、`query: [:keywords, :participant_id, :site_id]`の部分で対応できていると思っています。

`:page, :utf8, :commit,`の部分が不要、というのはよくわかっていません。
permitの役割は「viewから渡ってくるパラメータ以外（URLに手動で付与されたものなど）をはじく」だと思っていて、上記パラメータはviewから渡ってくるものなのでpermitするべき、という認識でした。

tricknotes replied:
> permitの役割は「viewから渡ってくるパラメータ以外（URLに手動で付与されたものなど）をはじく」だと思っていて、上記パラメータはviewから渡ってくるものなのでpermitするべき、という認識でした。

だいたい合ってるんですが、若干違います。
モデル(と限定した言い方をしてしまうと若干語弊がありますが…)が期待しているパラメータのみを許可するのが役割です。
そう考えると、モデルに不必要なパラメータを許可する理由はとくにないですよね。

myokoym replied:
なるほど、ありがとうございます。理解しました。
では、以下のように修正するのがよいでしょうか？
- メソッド名を`search_params`から`search_query_params`に変更する
- `:page, :utf8, :commit`を削除する

tricknotes replied:
いんじゃないでしょうか！

myokoym replied:
ありがとうございます！PR出します。

tricknotes commented on app/controllers/search_controller.rb:
```diff
@@ -0,0 +1,14 @@
+    @query = Query.new(search_params[:query])
+    @query.site_id = current_site.id
+    searcher = PostSearcher.new
+    @result_set = searcher.search_with_fallback_to_first_page(@query, page: search_params[:page])
```
ここは `params[:page]` でいいのでは

Commit:
test: close Groonga database to prevent a crash

tricknotes commented on test/test_helper.rb:
```diff
@@ -13,4 +13,29 @@
+
+  def teardown_groonga_database
+    context = Groonga::Context.default
+    database = context.database
```
ここの `context` が `nil` の状況ってありえないんですか？
数行下で  `Groonga::Context.default = nil` していたので、もしかして `nil` の場合があるのでは？という気持ちになりました。
(Groonga に詳しくない人の質問です)

myokoym replied:
たしかにまっさらな気持ちで見るとそんな気がしますね。。。
実際には以下のような挙動になっていて、

> デフォルトのコンテキストを設定する。 +nil+ を指定すると、 デフォルトのコンテキストをリセットする。リセットすると、次 回 default を呼び出したときに新しくコ ンテキストが作成される。
> http://ranguba.org/rroonga/ja/Groonga/Context.html#default%3D-class_method

この場合は直前で`Groonga::Context.default`を呼んでいるので大丈夫です！

tricknotes replied:
なるほど！勉強になります

tricknotes commented on app/views/search/_form.html.haml:
```diff
@@ -0,0 +1,3 @@
+= form_tag search_path, method: 'get' do
```
`'get'` は String (自由に記述できる文字)ではなくて Symbol (予約された単語のひとつ)の方がわたしのメンタルモデルにあっているんですが、あえて String を使いたい気持ちがあったりしますか？

myokoym replied:
ありません！シンボルにしますー

Commit:
test: remove a needless assignment

tricknotes commented on app/views/search/search.html.haml:
```diff
@@ -0,0 +1,28 @@
+  %ol
+    - @result_set.posts.each do |post|
+      %li
+        = link_to post.title, post_url(post.public_id), data: {score: post.score}
```
`Post#to_param` をオーバライドしているので、ここは `post_url(post.public_id)` じゃなくて `post` もしくは `post_url(post)` でよかったりしますか？

kou replied:
この`post`は`Post`ではなくてGroongaのレコードオブジェクトだと思います！
なので、`post`とか`post_url(post)`だとダメな気がします。

tricknotes replied:
あぁー、なるほど！
理解しました…！

AR の post と Groonga の post をパッと見て区別するには、 `result_set` から返ってきているかどうかで判別するのがよいのでしょうか…？

kou replied:
私が作っていたときは、検索処理の中にActive Recordのオブジェクトは登場しないようにしていました。

なので、データを更新するとき以外にでてくるとき以外はすべてGroongaのオブジェクト、と区別できます。できますというか、だった頃がありますが、今もそうなのかはわからないです。。。

tricknotes replied:
なるほど、ありがとうございます！
その気持ちになって（今との差分を気にしつつ）コードを見ることにしてみます！

Commit:
Add a test for only posts of the current site

tricknotes commented on test/test_helper.rb:
```diff
@@ -13,4 +13,24 @@
   # Add more helper methods to be used by all tests here...
   include FactoryGirl::Syntax::Methods
+
+  def setup_groonga_database
```
今のコードベースでは、ヘルパー系のメソッドは `test/support/` 以下に関心事毎にモジュールに分けておいてあるようですね！

myokoym replied:
そっちのほうがかっこいいですね…
分けました！

tricknotes commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,40 @@
+    click_on 'search'
+
+    within('main') do
+      assert_equal '「contents」を含む記事は1件見つかりました。', find('p').text
```
あとは、`@current_site` の記事であることを検証できるとより正確なテストになりそうですね！

myokoym replied:
ここで記事の件数が1件であることを検証して、`assert_equal 'current site', find('a').text`でその1件がcurrent siteの記事であることを検証しているつもりでした。もっとよさそうなやり方がありそうでしょうか…？

tricknotes replied:
なるほど…! `assert_equal 'current site', find('a').text` は記事の検証だったんですね。
記事のタイトルがサイト名っぽくなってしまっているので、そこが直れば問題ないかとー

myokoym replied:
たしかに…！直しました！

tricknotes commented:
見つかった記事が 50件(1ページの数)を超えた状態で 2ページ目を表示しても、記事の先頭についている番号が 1 からの連番になっていました。これは51からになっているのが正しそうですか？

tricknotes commented:
検索結果のページで、検索フォームに検索クエリを残さないようにしたのってなんでなんでしたっけ？

kou commented:
（最新のビューは見ていないので間違っていたらアレなんですけど、）検索結果のページでもまだサイドバーに検索フォームがあります？検索結果のページはサイドバーじゃなくて検索結果のところに検索フォームがあった方がユーザーが使いやすいんじゃないかなぁと思っています。

kou commented:
> 見つかった記事が 50件(1ページの数)を超えた状態で 2ページ目を表示しても、記事の先頭についている番号が 1 からの連番になっていました。これは51からになっているのが正しそうですか？

ページネーションをやめてもいいかもしれないですね。
50件以上見ることはないんじゃないですかねぃ。

Commit:
test: extract Groonga helper

Commit:
Use symbol for a constant value

Commit:
test: use Hash for arguments

tricknotes commented:
> 検索結果のページでもまだサイドバーに検索フォームがあります？

サイドバーにありました！

> 検索結果のページはサイドバーじゃなくて検索結果のところに検索フォームがあった方がユーザーが使いやすいんじゃないかなぁと思っています。

なるほど！そのほうがスムーズに再検索できそうですね！ :rocket:

Commit:
test: use more meaningful title of posts

Commit:
Don't use strong parameter if unnecessary

Commit:
test: fold long lines

Commit:
Add a test for not published post

Commit:
test: use "find_link"

Commit:
Tranalate search button label to Japanese

Commit:
Remove needless index

Words table is for only related posts search.

Commit:
Follow label language change

Commit:
Use Post model instead of Groonga record in search view

To use current updated_at to select published post.

@result_set.posts may contain stopped publishing post because
Groonga's database is updated at once a day for at least on
Heroku.

Note: This change breaks sorting by score.

Commit:
Remove needless "if" statement

tricknotes commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,77 @@
+    create_post(site: @current_site,
+                title: 'post2 is not published',
+                body: 'body...',
+                published_at: Time.now.since(1.hour))
```
`1.hour.from_now`

myokoym replied:
Railsish!

tricknotes commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,77 @@
+    end
+  end
+
+  test 'no published posts should be hide' do
```
- no published posts -> unpublished posts
- be hide -> be hidden 
  - 「隠れている」ではなく「表示されない」に主眼があるのだとしたら should not be shown

わたし、えいごにあまりじしんないです

myokoym replied:
ありがとうございます！とても参考になります！

Commit:
Prevent N+1

Commit:
test: fix English

tricknotes commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,77 @@
+    teardown_groonga_database
+  end
+
+  test 'only posts of the current site' do
```
ひとつ下のテストケース名がテスト対象と期待する結果を表現しているのに対して、こちらはテスト対象しか表現されていないですね…！
どうなっていてほしいか、を書いておくとテストをメンテナンスする人にとってこれが何のテストか理解しやすくて嬉しいんじゃないかと思いました。

myokoym replied:
ありがとうございます！その通りだと思いました！

> テストをメンテナンスする人にとって

心に刻んでおきます！

Commit:
test: indent

Commit:
test: use ActiveSupport::TimeWithZone to simplify

Commit:
test: fix English

Commit:
test: fix test case name

tricknotes commented on app/views/search/search.html.haml:
```diff
@@ -0,0 +1,27 @@
+    絞り込み：
+    - if @query.participant_id.present?
+      = Participant.find(@query.participant_id).name
+      = link_to content_tag(:span, '', class: 'fa fa-remove'), @query.params(remove: [:participant_id])
```
ここって、 `span` タグ使う必要ってあるんでしょうか？
わたしなら、[Font Awesome のサンプル](http://fortawesome.github.io/Font-Awesome/icon/times/)に書いてあるように `i` タグを使ってシンプルに書くなと思いました。

``` haml
= link_to @query.params(remove: [:participant_id]) do
  %i.fa.fa-remove
```

tricknotes replied:
すみません、ここでは２つの話が混ざってましたね…
- `content_tag` 使わなくても `link_to do` を使うほうがシンプルそう
- 特に理由がなければ `span` ではなく `i` を使ったほうが疑問を挟む余地が少ない

ダラっとコメントしてしまってすみませんでした！

myokoym replied:
丁寧なコメントありがとうございます！

> - `content_tag` 使わなくても `link_to do` を使うほうがシンプルそう

たしかにそちらのほうがHamlを活かせそうですね。
特にこだわりはないので `link_to do` にします。

myokoym replied:
> - 特に理由がなければ `span` ではなく `i` を使ったほうが疑問を挟む余地が少ない

意味的に正しいのは`span`だと思うのですが、Font Awesome的には`i`が使われることが多そうなので`i`にします。

myokoym replied:
メモ:

> You can place Font Awesome icons just about anywhere using the CSS Prefix fa and the icon's name. Font Awesome is designed to be used with inline elements (we like the `<i>` tag for brevity, but using a `<span>` is more semantically correct). 
> https://fortawesome.github.io/Font-Awesome/examples/

tricknotes replied:
hmhm

tricknotes commented on app/views/search/search.html.haml:
```diff
@@ -0,0 +1,27 @@
+      - @result_set.participants.each do |participant|
+        %span
+          = link_to participant.name, @query.params(add: {participant_id: participant._key})
+          = "(#{participant.n_sub_records})"
```
``` haml
= "#{...}"
```

をしたくなったら、haml の埋め込み記法を使うとスッキリしますね

``` haml
(#{participant.n_sub_records})
```

myokoym replied:
これは、`=`がない場所にRubyのコードを埋め込める、ということでしょうか。
そんな記法があったんですね！

myokoym replied:
と思ったら別の場所でも使っていましたね。要素がある場所とない場所で受ける印象が違うのかな…。
しかしこの部分は`()`が特別な意味を持っていそうに見えてしまいそうな気がします。うーん。

tricknotes replied:
洗練されたシンタックスハイライト機能をもつエディタで表示すると、そんなに気にならないと思いますよー

tricknotes replied:
ここはこれからなおしますか？

tricknotes commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,77 @@
+    create_post(site: @current_site,
+                title: 'post3 is published',
+                body: 'body...')
+    stopped_publishing_post = create_post(site: @current_site,
```
ここは unpublished ではなく stopped_publishing なのは何か強い思いがあるんでしょうか？
(新しい語彙が出てくると、コードを読んでいるひとはその言葉にはなにか意味があるのかなという気持ちで考えてしまいそうですね…！)

myokoym replied:
一度公開されてから非公開になったことを表現したかったのでした。

cosmo0920 replied:
よくわかってないですけど、temporary unpublished postとかonce the post in privateとかを使えば一時的な非公開ポストのような意味の変数名が出来そうな気がします。

tricknotes replied:
おぉ、なるほど！それはとってもわかりやすいと思います！

tricknotes replied:
これはここから(ry

tricknotes commented on test/integration/search_test.rb:
```diff
@@ -0,0 +1,77 @@
+    stopped_publishing_post = create_post(site: @current_site,
+                                          title: 'post4 is stopped publishing',
+                                          body: 'body...')
+    stopped_publishing_post.published_at = nil
```
`create_post` の引数には渡せませんでしたか？
see also https://github.com/bm-sms/daimon-news-multi-tenant/commit/247ba0ccecbe2c0cb01fed42d6fbd98f025df64c#commitcomment-16561313

myokoym replied:
一度公開されてから非公開にするケースをテストしたかったのでした。
テストを分けて名前を付ければわかりやすくなりますか。

tricknotes replied:
そうですね！
「一度公開されてから非公開にするケースをテストしたかったのでした。」というのがコードから読み取るのは難しかったので、せめてコードコメントがついている程度でも十分コードを読む手助けになると思いました！

tricknotes replied:
コメントはこれからつけますか？

Commit:
Use block style instead of content_tag

Commit:
Use <i> instead of <span> because Font Awesome likes <i>

tricknotes commented:
やはり、次のあたりが気になりました…(というのはわたしだけなのかな…。他のひとはどうなんでしょか)
* 検索後に、検索クエリが検索欄に残らないのがもやっとします
* 検索結果 2ページ目以降の連番が1始まりなのがもやっとします

これらについてはそれぞれにいくかの改善案が提示されていると思っています。
この PR でどこまでやるかという議論はありますが、これらの「もやっと」が解消されないとわたしは merge ready ではないと感じてます。(他のひとがどう思っているかはわからないですが。。。)

うーん、どうでしょうかねぇ…？

mtsmfm commented:
- 滞空が長い
- ひとまずユーザ影響はない

のでとりあえずマージしちゃいますか

tricknotes commented:
> とりあえずマージしちゃいますか

agreed :wink:

cosmo0920 commented:
:clap:

okkez commented:
:cake:
